infrastructure/src/main/java/org/corfudb/infrastructure/orchestrator/actions/RestoreRedundancyMergeSegments.java

```diff
@@ -1,38 +1,217 @@
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+
+import static org.corfudb.infrastructure.log.statetransfer.StateTransferManager.TransferSegmentStatus.SegmentState.FAILED;
+import static org.corfudb.infrastructure.log.statetransfer.StateTransferManager.TransferSegmentStatus.SegmentState.TRANSFERRED;
 
 /**
- * This action attempts to restore redundancy for all servers across all segments
- * starting from the oldest segment. It then collapses the segments once the set of
- * servers in the 2 oldest subsequent segments are equal.
- * Created by Zeeshan on 2019-02-06.
+ * This action attempts to restore the redundancy for the segments, for which the current node is
+ * missing. It transfers multiple segments at once and opens the segments that have been transferred.
  */
+@Slf4j
+@Getter
+@Builder
 public class RestoreRedundancyMergeSegments extends Action {
 
+    @Getter
+    @NonNull
+    private final String currentNode;
+
+    @Getter
+    @NonNull
+    private final RedundancyCalculator redundancyCalculator;
+
+    @Default
+    private final int retryBase = 3;
+
+    @Default
+    private final Duration backoffDuration = Duration.ofSeconds(10);
+
+    @Default
+    private final Duration extraWait = Duration.ofMillis(20);
+
+    @Default
+    private final float randomPart = 0.5f;
+
+    @Default
+    private final int restoreRetries = 3;
+
     /**
-     * Returns set of nodes which are present in the next index but not in the specified segment. These
-     * nodes have reduced redundancy and state needs to be transferred only to these before these segments can be
-     * merged.
+     * Perform a state transfer on a current node, if needed, and then
+     * propose a new layout based on a transfer result.
+     * If a state transfer was not needed, try merging the segments
+     * of a current layout and then proposing it.
+     * We utilize an exponential backoff since there can be cases
+     * when multiple nodes are proposing a new layout simultaneously.
      *
-     * @param layout             Current layout.
-     * @param layoutSegmentIndex Segment to compare to get nodes with reduced redundancy.
-     * @return Set of nodes with reduced redundancy.
+     * @param runtime         A corfu runtime.
+     * @param transferManager A transfer manager that runs the state transfer.
+     * @return A new layout, if a redundancy restoration occurred; a current layout otherwise.
      */
-    private Set<String> getNodesWithReducedRedundancy(Layout layout, int layoutSegmentIndex) {
-        // Get the set of servers present in the next segment but not in the this
-        // segment.
-        return Sets.difference(
-                layout.getSegments().get(layoutSegmentIndex + 1).getAllLogServers(),
-                layout.getSegments().get(layoutSegmentIndex).getAllLogServers());
+    Layout restoreWithBackOff(CorfuRuntime runtime, StateTransferManager transferManager)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use explicit scoping instead of the default package private level](https://app.codacy.com/app/annymartinez.m/CorfuDB/pullRequest?prid=4651453)